workspaces/optic-engine/src/projections/shape.rs

```diff
@@ -175,81 +175,93 @@ impl From<&ShapeProjection> for ShapeChoiceMapping {
-              ShapeKind::ObjectKind => {
-                let object_fields = queries.resolve_shape_field_id_and_names(&choice.shape_id);
-                let fields = object_fields
-                  .map(|(field_id, name)| {
-                    let field_shape_id = queries
-                      .resolve_field_shape_node(field_id)
-                      .expect("expected field shape to resolve");
-
-                    ObjectFieldChoice {
-                      field_id: field_id.clone(),
-                      shape_id: field_shape_id,
-                      name: name.clone(),
-                    }
-                  })
-                  .collect();
-
-                let output = ObjectChoice {
-                  shape_id: choice.shape_id.clone(),
-                  json_type: JsonType::Object,
-                  fields,
-                };
-                ShapeChoice::Object(output)
-              }
-              ShapeKind::ListKind => {
-                let shape_parameter_id = &String::from(
-                  choice
-                    .core_shape_kind
-                    .get_parameter_descriptor()
-                    .expect("expected $list to have a parameter descriptor")
-                    .shape_parameter_id,
-                );
-                let list_item_shape_id =
-                  queries.resolve_parameter_to_shape(&choice.shape_id, shape_parameter_id);
-                let output = ArrayChoice {
-                  shape_id: choice.shape_id.clone(),
-                  json_type: JsonType::Array,
-                  item_shape_id: list_item_shape_id,
-                };
-                ShapeChoice::Array(output)
-              }
-              ShapeKind::MapKind => unimplemented!(),
-              ShapeKind::OneOfKind => unreachable!(),
-              ShapeKind::AnyKind => ShapeChoice::Any,
-              ShapeKind::UnknownKind => ShapeChoice::Unknown,
-              ShapeKind::StringKind => Primitive(PrimitiveChoice {
-                shape_id: choice.shape_id.clone(),
-                json_type: JsonType::String,
-              }),
-              ShapeKind::NumberKind => Primitive(PrimitiveChoice {
-                shape_id: choice.shape_id.clone(),
-                json_type: JsonType::Number,
-              }),
-              ShapeKind::BooleanKind => Primitive(PrimitiveChoice {
-                shape_id: choice.shape_id.clone(),
-                json_type: JsonType::Boolean,
-              }),
-              ShapeKind::NullableKind => Primitive(PrimitiveChoice {
+          .find_map(|choice| match choice.parent_trail.last_component() {
+            Some(ShapeTrailPathComponent::OptionalItemTrail { shape_id, .. }) => {
+              Some(shape_id.clone())
+            }
+            _ => None,
+          })
+          .map(|optional_shape_id| {
+            Primitive(PrimitiveChoice {
+              shape_id: optional_shape_id,
+              json_type: JsonType::Undefined,
+            })
+          });
```

Comment:
Instead of mapping from choices 1:1, we generate the primitive `undefined` choice from looking at the parent trail, generating one when `OptionalItemTrail` is the last component of the shape trail. This passes the existing choice mappings snapshot tests.